src/pathfinder/opf/opf_ship.cpp

```diff
@@ -133,15 +139,16 @@ static const DiagDirection _ship_search_directions[6][4] = {
 /** Track to "direction (& 3)" mapping. */
 static const byte _pick_shiptrack_table[6] = {DIR_NE, DIR_SE, DIR_E, DIR_E, DIR_N, DIR_N};
 
-static uint FindShipTrack(const Ship *v, TileIndex tile, DiagDirection dir, TrackBits bits, TileIndex skiptile, Track *track)
+static uint FindShipTrack(const Ship *v, TileIndex tile, DiagDirection dir, TrackBits bits, TileIndex skiptile, Track *track, uint max_penalty = OPF_MAX_LENGTH, TileIndex depottile = INVALID_TILE)
 {
 	TrackPathFinder pfs;
 	uint best_bird_dist = 0;
 	uint best_length    = 0;
 	byte ship_dir = v->direction & 3;
 
-	pfs.dest_coords = v->dest_tile;
+	pfs.dest_coords = depottile != INVALID_TILE ? depottile : v->dest_tile;
 	pfs.skiptile = skiptile;
+	pfs.max_length = (max_penalty == 0 || max_penalty > OPF_MAX_LENGTH) ? OPF_MAX_LENGTH : max_penalty;
```

Comment:
Why the condition check for `max_penalty == 0`, if it's wrong to call the function with that parameter zero then make it an assert that it much be greater than zero, and change this line to a simple `max()` call.